uc-crux-llvm/src/UCCrux/LLVM/Classify/Poison.hs

```diff
@@ -0,0 +1,171 @@
+{-
+Module       : UCCrux.LLVM.Classify.Poison
+Description  : Classify errors as true positives or due to missing preconditions
+Copyright    : (c) Galois, Inc 2021
+License      : BSD3
+Maintainer   : Langston Barrett <langston@galois.com>
+Stability    : provisional
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE ViewPatterns #-}
+-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeOperators #-}
+
+module UCCrux.LLVM.Classify.Poison (classifyPoison) where
+
+{- ORMOLU_DISABLE -}
+import           Prelude hiding (log)
+
+import           Control.Lens ((^.), to)
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+
+import           Data.BitVector.Sized (BV)
+import qualified Data.BitVector.Sized as BV
+import           Data.Functor.Const (Const(getConst))
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+
+import qualified Text.LLVM.AST as L
+
+import           Data.Parameterized.Classes (IxedF'(ixF'))
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.NatRepr (NatRepr, type (<=))
+import           Data.Parameterized.Some (Some(Some))
+
+import qualified What4.Concrete as What4
+import qualified What4.Interface as What4
+
+import qualified Lang.Crucible.Simulator as Crucible
+
+import qualified Lang.Crucible.LLVM.Errors.Poison as Poison
+
+import           UCCrux.LLVM.Context.App (AppContext, log)
+import           UCCrux.LLVM.Context.Function (FunctionContext, argumentNames)
+import           UCCrux.LLVM.Classify.Types
+import           UCCrux.LLVM.Cursor (ppCursor, Selector(..), SomeInSelector(SomeInSelector))
+import           UCCrux.LLVM.Constraints (Constraint(BVCmp), NewConstraint(..))
+import           UCCrux.LLVM.FullType.Type (FullType(FTInt), FullTypeRepr(FTIntRepr))
+import           UCCrux.LLVM.Logging (Verbosity(Hi))
+import           UCCrux.LLVM.Setup.Monad (TypedSelector(..))
+{- ORMOLU_ENABLE -}
+
+getTermAnn ::
+  What4.IsExprBuilder sym =>
+  sym ->
+  -- | Term annotations (origins)
```

Comment:
Good point, that's vague - So this is a map from annotations to `Selector`s. Selectors point to a spot in a global variable or an argument, ie those values that are generated in Setup. This is how we know what value to attach a deduced precondition to. If the error is an OOB write and we look up the pointer in this map and it appears in an argument, we know to attach a precondition that that part of the argument gets allocated with more space. If we look up the pointer and it's not in this map, it's not clear what we could do to prevent it, and the error will be unclassified.

Reply:
Commented wherever this map appears!